waku/v2/peermanager/peer_manager.go

```diff
@@ -483,3 +498,140 @@ func (pm *PeerManager) selectServicePeer(proto protocol.ID, pubSubTopic string,
 	}
 	return
 }
+
+// PeerSelectionCriteria is the selection Criteria that is used by PeerManager to select peers.
+type PeerSelectionCriteria struct {
+	SelectionType PeerSelection
+	Proto         protocol.ID
+	PubsubTopic   string
+	SpecificPeers peer.IDSlice
+	Ctx           context.Context
+}
+
+// SelectPeer selects a peer based on selectionType specified.
+// Context is required only in case of selectionType set to LowestRTT
+func (pm *PeerManager) SelectPeer(criteria PeerSelectionCriteria) (peer.ID, error) {
+
+	switch criteria.SelectionType {
+	case Automatic:
+		return pm.SelectRandomPeer(criteria)
+	case LowestRTT:
+		if criteria.Ctx == nil {
+			criteria.Ctx = context.Background()
+			pm.logger.Warn("context is not passed for peerSelectionwithRTT, using background context")
+		}
+		return pm.SelectPeerWithLowestRTT(criteria)
+	default:
+		return "", errors.New("unknown peer selection type specified")
+	}
+}
+
+type pingResult struct {
+	p   peer.ID
+	rtt time.Duration
+}
+
+// SelectPeerWithLowestRTT will select a peer that supports a specific protocol with the lowest reply time
+// If a list of specific peers is passed, the peer will be chosen from that list assuming
+// it supports the chosen protocol, otherwise it will chose a peer from the node peerstore
+// TO OPTIMIZE: As of now the peer with lowest RTT is identified when select is called, this should be optimized
+// to maintain the RTT as part of peer-scoring and just select based on that.
+func (pm *PeerManager) SelectPeerWithLowestRTT(criteria PeerSelectionCriteria) (peer.ID, error) {
+	var peers peer.IDSlice
+	var err error
+	if criteria.Ctx == nil {
+		criteria.Ctx = context.Background()
+	}
+
+	if criteria.PubsubTopic != "" {
+		peers = pm.host.Peerstore().(wps.WakuPeerstore).PeersByPubSubTopic(criteria.PubsubTopic, criteria.SpecificPeers...)
+	}
+
+	peers, err = pm.FilterPeersByProto(peers, criteria.Proto)
+	if err != nil {
+		return "", err
+	}
+	wg := sync.WaitGroup{}
+	waitCh := make(chan struct{})
+	pingCh := make(chan pingResult, 1000)
+
+	wg.Add(len(peers))
+
+	go func() {
+		for _, p := range peers {
+			go func(p peer.ID) {
+				defer wg.Done()
+				ctx, cancel := context.WithTimeout(criteria.Ctx, 3*time.Second)
+				defer cancel()
+				result := <-ping.Ping(ctx, pm.host, p)
+				if result.Error == nil {
+					pingCh <- pingResult{
+						p:   p,
+						rtt: result.RTT,
+					}
+				} else {
+					fmt.Println("Error in Ping", result)
```

Comment:
```suggestion
					pm.logger.Error("could not ping", logging.HostID("peer", p), zap.Error(result.Error))
```

Do consider changing it to `Debug` tho, since there is going to be a lot of error logs depending on the number of peers, and failing a ping should be pretty common